### PR TITLE
REF: separate casting out of Index.__new__

### PR DESCRIPTION
**Pull request by jbrockmendel**
first of two PRs to separate array casting/inference out of `Index.__new__`.  Once both are in place, we'll be able to do all inference/casting up-front and simplify the constructor quite a bit.  We'll also be able to look into sharing code between Index/Series/array, and address a handful of outstanding issues with the Index constructor.

### CODE REVIEW

File: pandas/core/indexes/base.py [hunk 1]
```diff
@@ -5486,3 +5453,101 @@ def maybe_extract_name(name, obj, cls) -> Optional[Hashable]:
         raise TypeError(f"{cls.__name__}.name must be a hashable type")
 
     return name
+
+
+def _maybe_cast_with_dtype(data: np.ndarray, dtype: np.dtype, copy: bool) -> np.ndarray:
```
Comment (jreback) [inline pandas/core/indexes/base.py]: private?

Comment (jbrockmendel) [reply pandas/core/indexes/base.py]: until we have a reason to unprivatize it, i figure its safer this way

File: pandas/core/indexes/base.py [hunk 2]
```diff
@@ -5486,3 +5453,101 @@ def maybe_extract_name(name, obj, cls) -> Optional[Hashable]:
...
+    ValueError if the conversion was not successful.
+    """
+
+    if not is_unsigned_integer_dtype(dtype):
+        # skip int64 conversion attempt if uint-like dtype is passed, as
+        # this could return Int64Index when UInt64Index is what's desired
+        try:
+            res = data.astype("i8", copy=False)
+            if (res == data).all():
+                return res  # TODO: might still need to copy
```
Comment (jreback) [inline pandas/core/indexes/base.py]: should be in pandas/core/dtypes/cast.py and i think duplicates some code there, but ok for now.

File: pandas/core/indexes/base.py [hunk 3]
```diff
@@ -5486,3 +5453,101 @@ def maybe_extract_name(name, obj, cls) -> Optional[Hashable]:
...
+            pass
+        else:
+            data = data.astype(dtype)
+    else:
+        data = np.array(data, dtype=dtype, copy=copy)
+
+    return data
+
+
+def _try_convert_to_int_array(
```
Comment (jreback) [inline pandas/core/indexes/base.py]: you need to remove the original no?

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/indexes/base.py (hunk 2 above)
Comment (jbrockmendel) [reply pandas/core/indexes/base.py]: yes, there's definitely some duplication to get rid of, but first there are some kinks to iron out, in particular #21311, #17246.

File: pandas/core/indexes/base.py (hunk 3 above)
Comment (jbrockmendel) [reply pandas/core/indexes/base.py]: The original is still used once.  The next PR in this sequence removes that usage.

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks

**Pull request merged**